### PR TITLE
fix: don't mark cluster consolidated if constrained by budgets

### PR DESCRIPTION
**Pull request by njtran**
Fixes #N/A 

**Description**
Fixes a bug where we would mark the cluster as consolidated if budgets caused Consolidation to not occur. We only want to mark a cluster as consolidated if budgets aren't restricting the candidate set. 

**How was this change tested?**
make presubmit

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/disruption/multinodeconsolidation.go [hunk 1]
```diff
@@ -62,12 +62,15 @@ func (m *MultiNodeConsolidation) ComputeCommand(ctx context.Context, disruptionB
...
 	disruptableCandidates := make([]*Candidate, 0, len(candidates))
+	constrainedByBudgets := false
 	for _, candidate := range candidates {
 		// If there's disruptions allowed for the candidate's nodepool,
 		// add it to the list of candidates, and decrement the budget.
 		if disruptionBudgetMapping[candidate.nodePool.Name] == 0 {
 			continue
 		}
+		// set constrainedByBudgets to true if any node was a candidate but was constrained by a budget
+		constrainedByBudgets = true
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/multinodeconsolidation.go:73]: Does this make sense? Won't this always get called regardless if the candidate was constrained by a budget or not?

File: pkg/controllers/disruption/consolidation_test.go [hunk 2]
```diff
@@ -236,6 +238,8 @@ var _ = Describe("Consolidation", func() {
 			// Execute command, thus deleting all nodes
 			ExpectReconcileSucceeded(ctx, queue, types.NamespacedName{})
 			Expect(len(ExpectNodeClaims(ctx, env.Client))).To(Equal(numNodes))
+			// Expect the cluster consolidated state to not change.
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/consolidation_test.go:241]: sanity check: I'm assuming that this failed prior to this change?

Comment (jonathan-innis) [reply pkg/controllers/disruption/consolidation_test.go:241]: Should we test this in multiple places since we have 3 places where we are handling this logic but we only have a single test with emptiness where we are validating this behavior?

Comment (njtran) [reply pkg/controllers/disruption/consolidation_test.go:241]: The event is only emitted for consolidation when we have no allowed disruptions due to a blocking budget. I can add another test with multiple nodepools here

File: pkg/controllers/disruption/controller.go [hunk 3]
```diff
@@ -72,7 +72,7 @@ func NewController(clk clock.Clock, kubeClient client.Client, provisioner *provi
 		recorder:      recorder,
 		cloudProvider: cp,
 		lastRun:       map[string]time.Time{},
-		methods: []Method{
+		Methods: []Method{
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/controller.go:75]: Why the change to make methods publicly exposed?

File: pkg/controllers/disruption/consolidation_test.go [hunk 4]
```diff
@@ -227,6 +229,9 @@ var _ = Describe("Consolidation", func() {
 			ExpectReconcileSucceeded(ctx, disruptionController, client.ObjectKey{})
 			wg.Wait()
 
+			// Expect the cluster consolidated state to not change.
+			Expect(disruptionController.Methods[3].(*disruption.EmptyNodeConsolidation).IsConsolidated()).To(BeFalse())
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/consolidation_test.go:233]: This feels extremely dirty to me with exposing something publicly, getting a known ordering of it and then casting to get access to a member function. Is there a cleaner way to handle this that doesn't cause this many different hoops to jump through to get the testing to work-out?

File: pkg/controllers/disruption/multinodeconsolidation.go [hunk 5]
```diff
@@ -62,12 +62,15 @@ func (m *MultiNodeConsolidation) ComputeCommand(ctx context.Context, disruptionB
 	// would have violated the budget anyway, preserving the ordering
 	// and only considering a number of nodes that can be disrupted.
 	disruptableCandidates := make([]*Candidate, 0, len(candidates))
+	constrainedByBudgets := false
 	for _, candidate := range candidates {
 		// If there's disruptions allowed for the candidate's nodepool,
 		// add it to the list of candidates, and decrement the budget.
 		if disruptionBudgetMapping[candidate.nodePool.Name] == 0 {
+			constrainedByBudgets = true
```
Comment (Bryce-Soghigian) [inline pkg/controllers/disruption/multinodeconsolidation.go:70]: Constained by budgets might make for an interesting event to create on the cluster. We can use the DisruptionBlocked with the reaason being constrained by budgets?

File: pkg/controllers/disruption/consolidation.go [hunk 6]
```diff
@@ -88,8 +88,8 @@ func (c *consolidation) sortAndFilterCandidates(ctx context.Context, candidates
 	return candidates, nil
 }
 
-// isConsolidated returns true if nothing has changed since markConsolidated was called.
-func (c *consolidation) isConsolidated() bool {
+// IsConsolidated returns true if nothing has changed since markConsolidated was called.
```
Comment (Bryce-Soghigian) [inline pkg/controllers/disruption/consolidation.go:91]: Is there a need for this to be public for anything other than unit tests?

File: pkg/controllers/disruption/emptynodeconsolidation.go [hunk 7]
```diff
@@ -52,20 +52,29 @@ func (c *EmptyNodeConsolidation) ComputeCommand(ctx context.Context, disruptionB
...
 
 	empty := make([]*Candidate, 0, len(candidates))
+	constrainedByBudgets := false
 	for _, candidate := range candidates {
 		if len(candidate.pods) > 0 {
 			continue
 		}
+		if disruptionBudgetMapping[candidate.nodePool.Name] == 0 {
+			// set constrainedByBudgets to true if any node was a candidate but was constrained by a budget
+			constrainedByBudgets = true
```
Comment (Bryce-Soghigian) [inline pkg/controllers/disruption/emptynodeconsolidation.go:62]: If we ever can't remove a candidate due to a budget constraint, we should fire an event?

Comment (njtran) [reply pkg/controllers/disruption/emptynodeconsolidation.go:62]: We did something similar for NodePool here https://github.com/kubernetes-sigs/karpenter/pull/960. I think we can similarly add this for a candidate in the future too. One thing to consider is let's say a budget allows 3 disruptions, but we have 5 candidates. Do we emit an event for all 5 of the candidates, or just a random 2? This is probably trivial to do for Expiration and Drift, and we can just probably emit, but for multiNodeConsolidation, where the correctness of the command depends on how it's batched, this may make less sense.

File: pkg/controllers/disruption/consolidation_test.go [hunk 8]
```diff
@@ -461,6 +461,337 @@ var _ = Describe("Consolidation", func() {
...
+				})
+				Expect(found).To(BeTrue())
+				Expect(metric.GetGauge().GetValue()).To(BeNumerically("==", 0))
+			}
+
+			// Execute the command in the queue, deleting all node claims
+			ExpectReconcileSucceeded(ctx, queue, types.NamespacedName{})
+			Expect(len(ExpectNodeClaims(ctx, env.Client))).To(Equal(30))
+		})
+		It("should not mark empty node consolidated if candidates can't be disrupted due to budgets", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/consolidation_test.go:528]: Can you make this a bit more clear what the difference between this test and the one after it is? Maybe a context block or something to scope things appropriately if that's helpful. It was hard for me to quickly grok that these tests were relatively similar but had a one vs. many difference for the checks

File: pkg/controllers/disruption/consolidation_test.go [hunk 9]
```diff
@@ -461,6 +461,337 @@ var _ = Describe("Consolidation", func() {
...
+			var wg sync.WaitGroup
+			ExpectTriggerVerifyAction(&wg)
+			cmd, err := emptyConsolidation.ComputeCommand(ctx, budgets, candidates...)
+			Expect(err).To(Succeed())
+			Expect(cmd)
+			wg.Wait()
+
+			Expect(emptyConsolidation.IsConsolidated()).To(BeFalse())
+		})
+		It("should not mark multi node consolidated if candidates can't be disrupted due to budgets", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/consolidation_test.go:617]:
```suggestion
		It("should not mark multi node consolidated if candidates can't be disrupted due to budgets with one nodepool", func() {
```

File: pkg/controllers/disruption/consolidation_test.go [hunk 10]
```diff
@@ -461,6 +461,337 @@ var _ = Describe("Consolidation", func() {
...
+			var wg sync.WaitGroup
+			ExpectTriggerVerifyAction(&wg)
+			cmd, err := multiConsolidation.ComputeCommand(ctx, budgets, candidates...)
+			Expect(err).To(Succeed())
+			Expect(cmd)
+			wg.Wait()
+
+			Expect(multiConsolidation.IsConsolidated()).To(BeFalse())
+		})
+		It("should not mark multi node consolidated if all nodepools are blocked by budgets", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/consolidation_test.go:643]:
```suggestion
		It("should not mark multi node consolidated if all nodepools are blocked by budgets with many nodepools", func() {
```

File: pkg/controllers/disruption/consolidation_test.go [hunk 11]
```diff
@@ -461,6 +461,337 @@ var _ = Describe("Consolidation", func() {
...
+			var wg sync.WaitGroup
+			ExpectTriggerVerifyAction(&wg)
+			cmd, err := multiConsolidation.ComputeCommand(ctx, budgets, candidates...)
+			Expect(err).To(Succeed())
+			Expect(cmd)
+			wg.Wait()
+
+			Expect(multiConsolidation.IsConsolidated()).To(BeFalse())
+		})
+		It("should not mark single node consolidated if candidates can't be disrupted due to budgets", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/consolidation_test.go:706]:
```suggestion
		It("should not mark single node consolidated if candidates can't be disrupted due to budgets with one nodepool", func() {
```

File: pkg/controllers/disruption/consolidation_test.go [hunk 12]
```diff
@@ -461,6 +461,337 @@ var _ = Describe("Consolidation", func() {
...
+			var wg sync.WaitGroup
+			ExpectTriggerVerifyAction(&wg)
+			cmd, err := singleConsolidation.ComputeCommand(ctx, budgets, candidates...)
+			Expect(err).To(Succeed())
+			Expect(cmd)
+			wg.Wait()
+
+			Expect(singleConsolidation.IsConsolidated()).To(BeFalse())
+		})
+		It("should not mark single node consolidated if all nodepools are blocked by budgets", func() {
```
Comment (jonathan-innis) [inline pkg/controllers/disruption/consolidation_test.go:732]:
```suggestion
		It("should not mark single node consolidated if all nodepools are blocked by budgets with many nodepools", func() {
```

**Review by jonathan-innis — APPROVED** [review]
/lgtm
/approve
